### DISCUSSION TITLE
Compiling via C code

### DISCUSSION THREAD

**Discussion by WhoAteDaCake** [c-smile/sciter-sdk#208] [Q&A] [answered]
Hello, I've noticed that most headers are C compatible, but when I tried to compile, I encountered some issues like: 

> sciter-x-api.h:508:31: error: initializer element is not a compile-time constant

Is it possible to compile it via pure c, or are there features, which would not allow to do it ? The reason I'm asking is because I'm trying to create bindings to Crystal language. I'll try and see if I can apply fixes manually, but it would be good to know if it's not possible. Thanks !

**Comment by pravic** [comment]
In CGO it is patched as

https://github.com/sciter-sdk/go-sciter/blob/master/sciter-x-api.c#L10-L26

**Comment by c-smile** [comment]
Will this be better?

```C
  inline ISciterAPI* SAPI() {
#if defined(__cplusplus) && !defined(PLAIN_API_ONLY)
    static ISciterAPI* _api = _SAPI(NULL);
#else 
    static ISciterAPI* _api = 0;
    if (!_api) _api = _SAPI(NULL);
#endif
    return _api;
  }
```

**Reply by WhoAteDaCake** [reply]
Thanks for that, I've applied some changes, which can be seen in here: https://github.com/WhoAteDaCake/sciter-tests/blob/master/patches/0.patch

However, I'm still having a few issues when linking, I'm trying to call `SciterCreateWindow` function, which caused

>  undefined reference to `SciterCreateWindow'

Yet there was no error about missing `libsciter-gtk` .After running `objdump --demangle  -T libsciter-gtk.so` I found:

```
00000000003217d1 g    DF .text  000000000000031c  Base        SciterCreateWindow_api(unsigned int, tagRECT*, void**, void*, _GtkWidget*)
```
It seems like a linking error, but I can't seem to figure out the problem

**Comment by luxtau** [comment]
Did you compile this as WINDOWLESS?

https://github.com/c-smile/sciter-sdk/blob/29a598b6d20220b93848b5e8abab704619296857/include/sciter-x-api.h#L562-L565

**Reply by WhoAteDaCake** [reply]
No, I didn't specify any -D flags. My linker command looks something like:

> cc "${@}" -o /home/ubuntu/.cache/crystal/crystal-run-main.tmp  -rdynamic -L/usr/bin/../lib/crystal/lib -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lsciter-gtk -I/home/ubuntu/projects/sciter-tests/src/../include/ -L/home/ubuntu/projects/sciter-tests/src/../libs -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lpcre -lm -lgc -lpthread /usr/share/crystal/src/ext/libcrystal.a -levent -lrt -ldl

When fully extended

**Comment by WhoAteDaCake** [comment, marked as answer]
Thanks for the help. I managed to build it properly, by first building a C file, where I re-defined all functions, then use that library file as the main entry. You can see it here https://github.com/WhoAteDaCake/sciter-tests
